### PR TITLE
feat: restructure mock module and separate its codebase into a new crate

### LINKED ISSUE #951: [ICS24] Remove default value and implementation for `PortId`
## Problem Statement
In our `ics24_host` codebase implementation, it is not appropriate to have a default [value](https://github.com/cosmos/ibc-rs/blob/527bb141383129db804b0b9d49793c4a900deede/crates/ibc/src/core/ics24_host/identifier.rs#L18) and [implementation](https://github.com/cosmos/ibc-rs/blob/527bb141383129db804b0b9d49793c4a900deede/crates/ibc/src/core/ics24_host/identifier.rs#L419-L423) for the `PortId`. 
The sole purpose of this default value is for testing. It functions correctly for local unit tests, but when we traverse through the router, our router exclusively accepts the `transfer` `PortId` as valid. For this, even for testing purposes, it is better to utilize `transfer` value to ensure more comprehensive and meaningful test coverage.

## Proposal
Remove default value and implementation for `PortId`

### LINKED ISSUE #952: [ICS04] Expose channel message types as public
## Overview
Currently, our domain message types in the `ics04_channel` module are defined as private - `pub(crate)` - restricting external libraries from accessing and utilizing them. (At the moment, only these domain message types are not publicly accessible). 
This came to my attention when splitting our mock implementation into a separate crate. In order to craft IBC messages for various testing scenarios, the mock implementation requires access to these domain types.

## Proposal
Making the `ics04_channel` domain message types public to enable other libraries to access and use them as necessary.

### LINKED ISSUE #953: Separate `mock` module into a standalone crate
## Overview

As part of our efforts to make `ibc-rs` more streamlined, enhance usability, and easier to maintain, we are starting by separating the `mock` module into its own crate. This is a prerequisite step before [refactoring our mock testing facilities](https://github.com/cosmos/ibc-rs/issues/677), which ultimately will improve our testing coverage for ibc-rs and would simplify writing different testing scenarios by host chains. 

## Proposal
Create a separate crate for the mock module to achieve the following objectives:

1. Decouple the `ibc-rs` implementation from testing types and components.
2. Facilitate refactoring mock implementation so that can have more organized/comprehensive integration tests
3. Avoid circular dependencies, so would allow importing `basecoin-rs` libraries.
4. Make it easy to use the Merkle tree implementation within `basecoin-store`
5. Advance the goal of breaking down ibc-rs into smaller, self-contained and modular libraries

### PR DESCRIPTION
**Pull request by Farhad-Shabani**
Closes: #953
Closes: #952
Closes: #951

## Description

The mock implementation has undergone a comprehensive restructuring aimed at achieving complete decoupling of its various components and organizing them under relevant categories. This should facilitate future enhancements, particularly the addition of new implementations for components such as relayers and mock chains.

This restructuring results in four distinct top-level groups as follows:

- **hosts**: Encompasses code related to chains, with a particular focus on definitions related to the consensus layer of these chains.
- **testapp**: Codebases relevant to the application end (state transition machine) of a chain. It may contain various modules, including IBC.
- **relayer**: Houses types and functionalities required for simulating the role of a mock relayer, facilitating IBC talk between our testing chains.
- **utils**: Contains utility types, functions, and other tools that assist in writing tests.

By the way, this is the first iteration and we might improve it in the future as we refactor our testing kits.

The new crate is named `ibc-testkit`, and here is an overview of its directory structure:
```
.
├── hosts
│   ├── block.rs
│   └── mod.rs
├── lib.rs
├── relayer
│   ├── context.rs
│   ├── error.rs
│   └── mod.rs
├── testapp
│   ├── ibc
│   │   ├── applications
│   │   │   ├── mod.rs
│   │   │   └── transfer
│   │   ├── clients
│   │   │   ├── mock
│   │   │   └── mod.rs
│   │   ├── core
│   │   │   ├── client_ctx.rs
│   │   │   ├── core_ctx.rs
│   │   │   ├── mod.rs
│   │   │   ├── router
│   │   │   └── types.rs
│   │   └── mod.rs
│   └── mod.rs
└── utils
    ├── dummies
    │   ├── applications
    │   │   ├── mod.rs
    │   │   └── transfer.rs
    │   ├── clients
    │   │   ├── mock.rs
    │   │   ├── mod.rs
    │   │   └── tendermint.rs
    │   ├── core
    │   │   ├── channel
    │   │   ├── client
    │   │   ├── commitment.rs
    │   │   ├── connection
    │   │   ├── context.rs
    │   │   ├── mod.rs
    │   │   └── signer.rs
    │   ├── json
    │   │   ├── client_state.json
    │   │   ├── client_state_proof.json
    │   │   ├── consensus_state.json
    │   │   ├── consensus_state_proof.json
    │   │   └── signed_header.json
    │   └── mod.rs
    ├── fixture.rs
    └── mod.rs
```

### Remark
- The `mock` types are still maintained within the `ibc` crate.
- This PR also solves the issue reported here: https://github.com/cosmos/ibc-rs/issues/770. But keeping that issue open for adding something to CI as well.
______

### PR author checklist:
- [X] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [ ] Added tests.
- [X] Linked to GitHub issue.
- [X] Updated code comments and documentation (e.g., `docs/`).
- [X] Tagged *one* reviewer who will be the one responsible for shepherding this PR.

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

File: .changelog/unreleased/improvements/951-remove-default-portid.md [hunk 1]
```diff
@@ -0,0 +1,2 @@
+- Remove the default value and implementation for `PortId`
+  ([\#951](https://github.com/cosmos/ibc-rs/pull/951))
```
Comment (rnbguy) [inline .changelog/unreleased/improvements/951-remove-default-portid.md]:
```suggestion
  ([\#951](https://github.com/cosmos/ibc-rs/issues/951))
```

File: .changelog/unreleased/improvements/952-expose-channel-message-types copy.md [hunk 2]
```diff
@@ -0,0 +1,2 @@
+- Expose domain message types under the `ics04_channel` as public
+  ([\#952](https://github.com/cosmos/ibc-rs/pull/952))
```
Comment (rnbguy) [inline .changelog/unreleased/improvements/952-expose-channel-message-types copy.md]:
```suggestion
  ([\#952](https://github.com/cosmos/ibc-rs/issues/952))
```

File: .changelog/unreleased/features/954-restructure-mock-module-into-crate.md [hunk 3]
```diff
@@ -0,0 +1,3 @@
+- Restructure the mock module implementation and separate its codebase into a
+  new crate named `ibc-testkit`
+  ([\#954](https://github.com/cosmos/ibc-rs/issues/954))
```
Comment (rnbguy) [inline .changelog/unreleased/features/954-restructure-mock-module-into-crate.md]:
```suggestion
  ([\#954](https://github.com/cosmos/ibc-rs/issues/953))
```

File: README.md [hunk 4]
```diff
@@ -29,10 +29,11 @@ the `ibc` rust crate which defines the main data structures and on-chain logic f
 ## Libraries
 
 - [ibc](crates/ibc/README.md) - Data structures and on-chain logic for the IBC protocol.
-- [ibc-query](crates/ibc-query/README.md) - Utility traits and implementations for querying the
-state of an `ibc-rs` enabled chain.
 - [ibc-derive](crates/ibc-derive/README.md) - Derive macros for `ClientState`
   and `ConsensusState` traits, reducing boilerplate.
+- [ibc-testkit](crates/ibc-testkit/README.md) - Testing toolkit to aid ibc-rs and host chains for writing integration tests
```
Comment (rnbguy) [inline README.md]:
```suggestion
- [ibc-testkit](crates/ibc-testkit/README.md) - Testing toolkit to aid `ibc-rs` and host chains in writing integration tests.
```

File: crates/ibc-testkit/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+readme       = "README.md"
+keywords     = ["blockchain", "consensus", "cosmos", "ibc", "tendermint"]
+repository   = "https://github.com/cosmos/ibc-rs"
+authors      = ["Informal Systems <hello@informal.systems>"]
+rust-version = "1.60"
+description  = """
+    `ibc-testkit` is a versatile library that provides essential abstractions and implementations, 
+    fulfilling a dual role of enabling rigorous integration testing for the ibc-rs implementation 
+    while also aiding host chains in addressing a broad spectrum of testing scenarios during 
+    their integration with ibc-rs.
```
Comment (rnbguy) [inline crates/ibc-testkit/Cargo.toml]:
```suggestion
    `ibc-testkit` is a versatile library that provides essential abstractions and implementations, 
    fulfilling a dual role of enabling rigorous integration testing for the `ibc-rs` implementation 
    while also aiding host chains in addressing a broad spectrum of testing scenarios during 
    their integration with `ibc-rs`.
```

File: crates/ibc-testkit/Cargo.toml [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+]
+
+# This feature is required for token transfer (ICS-20)
+serde = ["dep:serde", "ibc/serde", "serde_json"]
+
+[dependencies]
+bytes = { version = "1.2.1", default-features = false }
+derive_more = { version = "0.99.17", default-features = false, features = ["from", "into", "display", "try_into"] }
+displaydoc = { version = "0.2", default-features = false }
+ibc = { verson = "0.47.0" , path = "../ibc", default-features = false, features = ["test-utils"] }
```
Comment (rnbguy) [inline crates/ibc-testkit/Cargo.toml]:
```suggestion
ibc = { version = "0.47.0" , path = "../ibc", default-features = false, features = ["test-utils"] }
```

File: crates/ibc-testkit/README.md [hunk 7]
```diff
@@ -0,0 +1,6 @@
+# `ibc-testkit`
+
+`ibc-testkit` is a versatile library that provides essential abstractions and
+implementations, fulfilling a dual role of enabling rigorous integration testing
+for the ibc-rs implementation while also aiding host chains in addressing a
+broad spectrum of testing scenarios during their integration with ibc-rs.
```
Comment (rnbguy) [inline crates/ibc-testkit/README.md]:
```suggestion
`ibc-testkit` is a versatile library that provides essential abstractions and
implementations, fulfilling a dual role of enabling rigorous integration testing
for the `ibc-rs` implementation while also aiding host chains in addressing a
broad spectrum of testing scenarios during their integration with `ibc-rs`.
```

File: crates/ibc/Cargo.toml [hunk 8]
```diff
@@ -44,9 +42,9 @@ serde = ["dep:serde", "dep:serde_derive", "serde_json", "ibc-proto/serde", "ics2
 # CosmWasm message API generator compatible, should not be inside on chain code
 schema = ["dep:schemars", "ibc-proto/json-schema", "serde", "std"]
 
-# This feature grants access to development-time mocking libraries, such as `MockContext` or `MockHeader`.
-# Depends on the `testgen` suite for generating Tendermint light blocks.
-mocks = ["tendermint-testgen", "parking_lot", "typed-builder", "std"]
+# This feature provides access to utility implementations for IBC types,
+# specifically designed for testing purposes
```
Comment (rnbguy) [inline crates/ibc/Cargo.toml]:
```suggestion
# specifically designed for testing purposes.
```

File: crates/ibc/src/clients/ics07_tendermint/header.rs [hunk 9]
```diff
@@ -249,7 +249,7 @@ mod pretty {
     }
 }
 
-#[cfg(any(test, feature = "mocks"))]
+#[cfg(any(test, feature = "test-utils"))]
 pub mod test_util {
```
Comment (rnbguy) [inline crates/ibc/src/clients/ics07_tendermint/header.rs]: Should we use `test_utils` as convention? I see we have `crate::test_utils` now.

```suggestion
pub mod test_utils {
```

File: crates/ibc/src/mock/header.rs [hunk 10]
```diff
@@ -70,15 +70,17 @@ impl MockHeader {
...
+            if #[cfg(any(test, feature = "std"))] {
+                Self {
+                    height,
+                    timestamp: Timestamp::now(),
+                }
+            } else {
+                Self {
+                    height,
+                    timestamp: Timestamp::none(),
+                }
```
Comment (rnbguy) [inline crates/ibc/src/mock/header.rs]: IMO we can make `MockHeader::new` return with `Timestamp::none()`. and add another method `with_current_timestamp()` with `#[cfg(any(test, feature = "std"))]`.

Comment (rnbguy) [reply crates/ibc/src/mock/header.rs]: btw, do we need `cfg_if`? [we can just do](https://doc.rust-lang.org/rust-by-example/attribute/cfg.html)
```
if cfg!(any(test, feature = "std)) {
    ...
} else { 
    ...
}
```

File: crates/ibc-testkit/src/testapp/ibc/core/client_ctx.rs [hunk 11]
```diff
@@ -29,7 +44,7 @@ impl MockClientContext for MockContext {
     fn consensus_state(
         &self,
         client_cons_state_path: &ClientConsensusStatePath,
-    ) -> Result<Self::AnyConsensusState, ContextError> {
+    ) -> Result<AnyConsensusState, ContextError> {
```
Comment (rnbguy) [inline crates/ibc-testkit/src/testapp/ibc/core/client_ctx.rs]: I would prefer to keep it as `Self::AnyConsensusState`.

File: crates/ibc-testkit/src/testapp/ibc/core/types.rs [hunk 12]
```diff
@@ -1709,7 +937,7 @@ mod tests {
     fn test_router() {
         #[derive(Debug, Default)]
         struct FooModule {
-            counter: u64,
+            counter: usize,
```
Comment (rnbguy) [inline crates/ibc-testkit/src/testapp/ibc/core/types.rs]: Why is this needed? Don't we prefer `u64` for cross compatiblity reasons?

**Review by rnbguy — CHANGES_REQUESTED** [review]
Thanks @Farhad-Shabani for taking care of this big PR. Just a few suggestions and questions.

File: crates/ibc/src/clients/ics07_tendermint/header.rs (hunk 9 above)
Comment (Farhad-Shabani) [reply crates/ibc/src/clients/ics07_tendermint/header.rs]: The codebase under `test_utils` is now under `ibc-testkit`, so these modules are removed.

File: crates/ibc/src/mock/header.rs (hunk 10 above)
Comment (Farhad-Shabani) [reply crates/ibc/src/mock/header.rs]: `cfg_if` was dropped since mock types moved to the `ibc-testkit` and `ibc-testkit` doesn't well support `no-std`

File: crates/ibc/src/clients/ics07_tendermint/client_state.rs [hunk 13]
```diff
@@ -792,19 +792,102 @@ fn check_header_trusted_next_validator_set(
     }
 }
 
+#[cfg(test)]
+pub mod test_utils {
```
Comment (rnbguy) [inline crates/ibc/src/clients/ics07_tendermint/client_state.rs]: is it being used anywhere else? if not, I would suggest to merge this block under `mod test { ... }`.

File: crates/ibc/src/clients/ics07_tendermint/consensus_state.rs [hunk 14]
```diff
@@ -172,14 +171,15 @@ mod tests {
     #[test]
     fn serialization_roundtrip_no_proof() {
         let json_data =
-            include_str!("../../../tests/support/query/serialization/consensus_state.json");
+            include_str!("../../../../ibc-testkit/src/utils/dummies/json/consensus_state.json");
```
Comment (rnbguy) [inline crates/ibc/src/clients/ics07_tendermint/consensus_state.rs]: is it okay to move the `json` files as following
```
ibc-testkit/src/data/json/dummy_consensus_state.json
```

also, I don't prefer relative path as this is relative to the current file - which makes the refactoring messy. can we use `concat!(env!("CARGO_MANIFEST_DIR"), "/../ibc-testkit/.....")`?

**Review by rnbguy — APPROVED** [review]
Just few more comments :slightly_smiling_face: Otherwise, I approved the PR.

This was a huge effort ! Thanks @Farhad-Shabani :pray:

File: crates/ibc/src/clients/ics07_tendermint/client_state.rs (hunk 13 above)
Comment (Farhad-Shabani) [reply crates/ibc/src/clients/ics07_tendermint/client_state.rs]: [e88452c](https://github.com/cosmos/ibc-rs/pull/954/commits/e88452cff23461b00ecf08ea60f4d11a67e89449)

File: crates/ibc/src/clients/ics07_tendermint/consensus_state.rs (hunk 14 above)
Comment (Farhad-Shabani) [reply crates/ibc/src/clients/ics07_tendermint/consensus_state.rs]: Good point: [15a01d1](https://github.com/cosmos/ibc-rs/pull/954/commits/15a01d17c219c68ecd4388577c5ae14e3ed78e1b)

Comment (rnbguy) [reply crates/ibc/src/clients/ics07_tendermint/consensus_state.rs]: On a second thought, it is better to keep the dummy JSON files under `ibc` project directly, as it is the only place it is being used. Anyway, we aren't exporting the dummy JSON files in `ibc-testkit`.

But keep them under `/ibc/tests/data/dummies/json/*.json`.

Comment (rnbguy) [reply crates/ibc/src/clients/ics07_tendermint/consensus_state.rs]: Also, we should use the term `fixture` instead of `dummy`. Maybe in future PR. :wink:

Comment (Farhad-Shabani) [reply crates/ibc/src/clients/ics07_tendermint/consensus_state.rs]: > Also, we should use the term fixture instead of dummy. Maybe in future PR. 😉

Those are the inherited naming conventions, yes. better to address in a separate PR.
